### PR TITLE
workaround for num_programs() to get correct value

### PR DESCRIPTION
**Pull request by scxiao**
This PR is a workaround for the function `tl.num_programs()` to return correct value. Reason for this problem is that mlir::gpu gridDimOp has a bug for gridDim to return the number of threads in a kernel (it should return the number of workgroups in the kernel), please see https://reviews.llvm.org/D156009.

So the workaround is to use grimDimOp/blockDimOp as the return value of `tl.num_programs()`. When llvm fix is available and integrated to triton, we should remove this workaround. A unit test is added to ensure that the workaround will be removed when we do upstream.

For issue:
https://github.com/ROCmSoftwarePlatform/frameworks-internal/issues/5016
https://github.com/ROCmSoftwarePlatform/triton/issues/259

### CODE REVIEW

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp [hunk 1]
```diff
@@ -464,16 +464,37 @@ struct GetNumProgramsOpConversion
...
+    // Seem like GridDimOp returns the number of threads (not the number of 
+    // workgroups) in a kernel (a bug in llvm https://reviews.llvm.org/D156009), 
+    // so as a workaround here, we divide by the number of threads 
+    // per workgroup to get the number of workgroups in a kernel.
+    // TODO: when we do upstream to include llvm fix, we can remove this workaround
+    // The unit test added in this PR can guarantee that.
+    Value threadsPerGrid =
+        rewriter.create<::mlir::gpu::GridDimOp>(loc, dims[op.getAxis()]);
+    Value threadsPerBlock =
+        rewriter.create<::mlir::gpu::BlockDimOp>(loc, bdims[op.getAxis()]);
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:477]: Why cannot we reuse `dims` in this case?

**Review by zhanglx13 — APPROVED** [review]
The introduction of `div` op will harm the performance.
But usually it is done at once and we do not use it for GEMM or FA. 
So approved.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 1 above)
Comment (scxiao) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:477]: My understanding is that dims is used to store the `gridDim` information, so we need another one to store the blockDim information.

Comment (zhanglx13) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:477]: `dims` is just a const array providing the three dimensions. And you do want `threadsPerGrid` and `threadsPerBlock` to have the same dim. So I think you can safely reuse `dims` and remove `bdims` below.

Comment (scxiao) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp]: Thanks. Fixed.

**Pull request merged**